lib/vecmat_external.h

```diff
@@ -43,118 +43,250 @@
 #pragma once
 
 #include "fix.h"
+#include <array>
+#include <numeric>
+#include <algorithm>
+#include <limits>
+
+template<typename T>
+constexpr int VM_BIT_SCAN_REVERSE_CONST(const T n) noexcept {
+    if (n == 0) return -1;
+    T a = n, b = 0, j = std::numeric_limits<T>::digits, k = 0;
+    do {
+        j >>= 1;
+        k = (T)1 << j;
+        if (a >= k) {
+            a >>= j;
+            b += j;
+        }
+    } while (j > 0);
+    return int(b);
+}
 
-struct angvec {
-using T = angle;
-constexpr static const size_t N       = 3;
-constexpr static const size_t PITCH   = 0;
-constexpr static const size_t HEADING = 1;
-constexpr static const size_t BANK    = 2;
-union {
-  T phb[N];
-  struct { T p, h, b; };
+template<typename T>
+constexpr inline T VM_BIT_CEIL(T x) noexcept { return T(1) << (VM_BIT_SCAN_REVERSE_CONST((T)x-1) + 1); };
+
+#ifdef _MSVC_LANG
+#define VM_ALIGN(n) __declspec(align(n))
+#else
+#define VM_ALIGN(n) __attribute__((aligned(n)))
+#endif
+
+enum class align
+{
+	none     = 0 << 0,
+	scalar   = 1 << 1,
+	vector   = 2 << 1,
+	matrix   = 3 << 1,
+	adaptive = 4 << 1,
 };
-constexpr static inline angvec id(ssize_t i = -1)
+
+template<typename T, size_t N, enum align A = align::adaptive, size_t N_POW2 = VM_BIT_CEIL(N)>
```

Comment:
not following why we're doing all of that, what was the issue with what we had before your latest changes?

Reply:
It allows types to be based on scalar/vector aligned std::array which allows use of [static extents/mdspan](https://en.cppreference.com/w/cpp/container/mdspan/extents/static_extent) and compatibility with other C++ container interfaces.

Reply:
Use of SIMD optimization for vector aligned types for -O3